exp/ingest/main.go

```diff
@@ -77,3 +82,71 @@ type Session interface {
 	// UpdateUnlock unlocks write lock of the session.
 	UpdateUnlock()
 }
+
+// StateReporter can be used by a session to log progress
+// or update metrics as the session runs its state pipelines.
+type StateReporter interface {
+	// OnStartState is called when the session begins processing
+	// a history archive snapshot at the given sequence number.
+	OnStartState(sequence uint32)
+	// OnStateEntry is called when the session processes an entry
+	// from the io.StateReader
+	OnStateEntry()
+	// OnEndState is called when the session finishes processing
+	// a history archive snapshot.
+	// if err is not nil it means that the session stoped processing the
+	// history archive snapshot because of an error.
+	// if shutdown is true it means the session stopped processing the
+	// history archive snapshot because it received a shutdown signal
+	OnEndState(err error, shutdown bool)
+}
+
+// reporterLedgerReader instruments a io.StateReader with a StateReporter
+// which reports each xdr.LedgerEntryChange which is read from the reader
+type reporterStateReader struct {
+	io.StateReader
+	StateReporter
+}
+
+func (r reporterStateReader) Read() (xdr.LedgerEntryChange, error) {
+	entry, err := r.StateReader.Read()
+	if err != nil {
```

Comment:
```suggestion
	if err == nil {
```
Currently it calls `OnStateEntry` only for errors.